sigtool/sigtool.c

```diff
@@ -326,6 +326,97 @@ static int hashsig(const struct optstruct *opts, unsigned int class, int type)
     return 0;
 }
 
+static int fuzzy_img_file(char *filename)
+{
+    int status = -1;
+
+    int target_fd                   = -1;
+    FFIError *fuzzy_hash_calc_error = NULL;
+    uint8_t *mem                    = NULL;
+
+    image_fuzzy_hash_t hash = {0};
+    STATBUF st;
+    ssize_t bytes_read;
+
+    if ((target_fd = open(filename, O_RDONLY)) == -1) {
+        mprintf(LOGG_ERROR, "%s: Can't open file\n", basename(filename));
+        goto done;
+    }
+
+    if (FSTAT(target_fd, &st)) {
+        mprintf(LOGG_ERROR, "%s: fstat() failed.\n", basename(filename));
+        goto done;
+    }
+
+    if (NULL == (mem = malloc((size_t)st.st_size))) {
+        mprintf(LOGG_ERROR, "%s: Malloc failed, buffer size: %zu\n", basename(filename), (size_t)st.st_size);
+        goto done;
+    }
+
+    bytes_read = read(target_fd, mem, (size_t)st.st_size);
+    if (bytes_read != (ssize_t)st.st_size) {
```

Comment:
Technically, there are two possible failure results from `read()`, `-1` if an error is encountered, or a smaller-than-expected read size (which should never happen unless the file descriptor is non-blocking). I would handle these separately.